flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/aggregate/RowTimeUnboundedOver.java

```diff
@@ -0,0 +1,266 @@
+
+		// input element are all binary row as they are came from network
+		BaseRowTypeInfo inputType = new BaseRowTypeInfo(inputFieldTypes);
+		ListTypeInfo<BaseRow> rowListTypeInfo = new ListTypeInfo<BaseRow>(inputType);
+		MapStateDescriptor<Long, List<BaseRow>> inputStateDesc = new MapStateDescriptor<Long, List<BaseRow>>(
+			"inputState",
+			Types.LONG,
+			rowListTypeInfo);
+		inputState = getRuntimeContext().getMapState(inputStateDesc);
+
+		initCleanupTimeState("RowTimeUnboundedOverCleanupTime");
+	}
+
+	/**
+	 * Puts an element from the input stream into state if it is not late.
+	 * Registers a timer for the next watermark.
+	 *
+	 * @param input The input value.
+	 * @param ctx   A {@link Context} that allows querying the timestamp of the element and getting
+	 *              a {@link TimerService} for registering timers and querying the time. The
+	 *              context is only valid during the invocation of this method, do not store it.
+	 * @param out   The collector for returning result values.
+	 * @throws Exception
+	 */
+	@Override
+	public void processElement(
+		BaseRow input,
+		KeyedProcessFunction<K, BaseRow, BaseRow>.Context ctx,
+		Collector<BaseRow> out) throws Exception {
+		// register state-cleanup timer
+		registerProcessingCleanupTimer(ctx, ctx.timerService().currentProcessingTime());
+
+		Long timestamp = input.getLong(rowTimeIdx);
+		Long curWatermark = ctx.timerService().currentWatermark();
+
+		// discard late record
+		if (timestamp > curWatermark) {
+			// ensure every key just registers one timer
+			// default watermark is Long.Min, avoid overflow we use zero when watermark < 0
+			Long triggerTs = curWatermark < 0 ? 0 : curWatermark + 1;
+			ctx.timerService().registerEventTimeTimer(triggerTs);
+
+			// put row into state
+			List<BaseRow> rowList = inputState.get(timestamp);
+			if (rowList == null) {
+				rowList = new ArrayList<BaseRow>();
+			}
+			rowList.add(input);
+			inputState.put(timestamp, rowList);
+		}
+	}
+
+	@Override
+	public void onTimer(
+		long timestamp,
+		KeyedProcessFunction<K, BaseRow, BaseRow>.OnTimerContext ctx,
+		Collector<BaseRow> out) throws Exception {
+		if (isProcessingTimeTimer(ctx)) {
+			if (needToCleanupState(timestamp)) {
+
+				// we check whether there are still records which have not been processed yet
+				Boolean noRecordsToProcess = !inputState.contains(timestamp);
+				if (noRecordsToProcess) {
+					// we clean the state
+					cleanupState(inputState, accState);
+					function.cleanup();
+				} else {
+					// There are records left to process because a watermark has not been received yet.
+					// This would only happen if the input stream has stopped. So we don't need to clean up.
+					// We leave the state as it is and schedule a new cleanup timer
+					registerProcessingCleanupTimer(ctx, ctx.timerService().currentProcessingTime());
+				}
+			}
+			return;
+		}
+
+		Iterator<Map.Entry<Long, List<BaseRow>>> keyIterator = inputState.iterator();
```

Comment:
Use `inputState.keys()` instead of `inputState.iterator()` to avoid unnecessary cost.